frontend/lib/src/components/elements/GraphVizChart/GraphVizChart.tsx

```diff
@@ -38,75 +38,68 @@ interface Dimensions {
+
 export function GraphVizChart({
   width: propWidth,
   element,
   height: propHeight,
 }: GraphVizChartProps): ReactElement {
   const chartId = `graphviz-chart-${element.elementId}`
 
+  const isFull = isFullScreen(propHeight)
+  const isFullRef = useRef(isFull)
+  // Update isFull state whenever propHeight changes
+  useEffect(() => {
+    isFullRef.current = isFull
+  }, [isFull])
+
   let originalHeight = 0
   let originalWidth = 0
 
-  const getChartData = (): string => {
-    return element.spec
-  }
+  const setSvgDimensions = useMemo(() => {
+    return (node: SVGGraphicsElement): void => {
+      const bbox = node.getBBox()
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      originalHeight = Math.round(bbox.height)
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      originalWidth = Math.round(bbox.width)
 
-  const getChartDimensions = (): Dimensions => {
-    let chartWidth = originalWidth
-    let chartHeight = originalHeight
-
-    if (propHeight) {
-      // fullscreen
-      chartWidth = propWidth
-      chartHeight = propHeight
-    } else if (element.useContainerWidth) {
-      chartWidth = propWidth
+      select(node)
+        .attr("width", isFullRef.current ? "100%" : `${originalWidth}pt`)
+        .attr("height", isFullRef.current ? "100%" : `${originalHeight}pt`)
     }
+  }, [isFullRef])
+
+  const getChartDimensions = (): Dimensions => {
+    const chartWidth =
+      isFullRef.current || element.useContainerWidth
+        ? propWidth
+        : originalWidth
+    const chartHeight = isFullRef.current
+      ? propHeight || originalHeight
+      : originalHeight
     return { chartWidth, chartHeight }
   }
 
-  const updateChart = (): void => {
+  useEffect(() => {
     try {
       // Layout and render the graph
-      const graph = graphviz(`#${chartId}`)
-        .zoom(false)
-        .fit(true)
-        .scale(1)
-        .renderDot(getChartData())
-        .on("end", () => {
-          const node = select(`#${chartId} > svg`).node() as SVGGraphicsElement
-          if (node) {
-            originalHeight = node.getBBox().height
-            originalWidth = node.getBBox().width
-          }
-        })
-
-      const { chartHeight, chartWidth } = getChartDimensions()
-      if (chartHeight > 0) {
-        // Override or reset the graph height
-        graph.height(chartHeight)
-      }
-      if (chartWidth > 0) {
-        // Override or reset the graph width
-        graph.width(chartWidth)
-      }
+      const graph = graphviz(`#${chartId}`).zoom(false).fit(true).scale(1)
+
+      graph.renderDot(element.spec).on("end", () => {
+        const node = select(`#${chartId} > svg`).node()
+        setSvgDimensions(node as SVGGraphicsElement)
+      })
     } catch (error) {
       logError(error)
     }
-  }
-
-  useEffect(() => {
-    updateChart()
-  })
+  }, [propHeight, propWidth, element.spec, chartId, setSvgDimensions])
```

Comment:
Previously we didn't include a dependency array in the `useEffect` so this would run each render - is there a reason we are doing so now?

Reply:
I included the dependency array now purely for performance and correctness reasons. When omitted, there's a very noticeable delay/lag between renders. When included, the renders are comparatively faster and we're ensuring the effect only runs when certain values (we know should trigger a render) have changed.

Lastly, @mayagbarnes I'm so sorry for taking up a lot of your time with these reviews! 😅 Feel like I should have shared a spec first before jumping to a fix.